File: week8/김정호/타겟 넘버.java
class Solution {
    public static void main(String[] args) {
        // 테스트 코드
        Solution solution = new Solution();

        int[] numbers = new int[]{1,1,1,1,1};
        int target = 3;

        System.out.println(solution.solution(numbers, target));
    }

    static int answer = 0;
    public int solution(int[] numbers, int target) {
        DFS(numbers, 0, 0, target);

        return answer;
    }

    public static void DFS(int[] numbers, int depth, int sum, int target) {
        if (depth == numbers.length) {
            if(sum == target) {
                answer++;
            }
            return;
        }

        DFS(numbers,depth + 1, sum + numbers[depth], target);
        DFS(numbers,depth + 1, sum - numbers[depth], target);
    }
}

File: week8/한지희/타겟_넘버.kt
class TargetNumber {
    var target = 0
    var answer = 0
    fun solution(numbers: IntArray, target: Int): Int {
        this.target = target
        dfs(numbers, 0, 0)

        return answer
    }

    private fun dfs(numbers: IntArray, depth: Int, sum: Int) {
        if (depth == numbers.size) {
            if (target == sum) answer++
            return
        }

        dfs(numbers, depth + 1, sum + numbers[depth])
        dfs(numbers, depth + 1, sum - numbers[depth])
    }

    fun bestSolution(numbers: IntArray, target: Int): Int {
        return numbers.fold(listOf(0)) { list, i ->
            list.run {
                map { it + i } + map { it - i }
            }
        }.count { it == target }
    }
}